source/device_defender.c

```diff
@@ -269,6 +301,90 @@ static int s_get_metric_report_json(
         goto cleanup;
     }
 
+    if (custom_metrics_len != 0) {
+        struct cJSON *custom_metrics = cJSON_CreateObject();
+        if (NULL == custom_metrics) {
+            goto cleanup;
+        }
+        cJSON_AddItemToObject(root, "custom_metrics", custom_metrics);
+
+        size_t list_size = 0;
+        struct cJSON *array_item = NULL;
+        struct cJSON *item = NULL;
+        struct cJSON *json_list = NULL;
+        struct cJSON *spurious_array_container = NULL;
+        for (size_t metric_index = 0; metric_index < custom_metrics_len; ++metric_index) {
+            spurious_array_container = cJSON_CreateArray();
+            if (NULL == spurious_array_container) {
+                goto cleanup;
+            }
+            cJSON_AddItemToObject(
+                custom_metrics,
+                aws_string_c_str(custom_metrics_data[metric_index].metric->metric_name),
+                spurious_array_container);
+
+            item = cJSON_CreateObject();
+            if (NULL == item) {
+                goto cleanup;
+            }
+            cJSON_AddItemToArray(spurious_array_container, item);
+
+            switch (custom_metrics_data[metric_index].metric->type) {
```

Comment:
Per slack thread, it feels like things would be greatly simplified with a json-oriented interface.

Reply:
Absolutely. I think there are far better styles of creating a tree structure that doesn't push so much potential for error checking on each stage of it's construction.